games/twenty_forty_eight/game.py

```diff
@@ -0,0 +1,76 @@
+            action = random.choice(self.actions_for(state, is_opponent=True))
+            state = self.take_action(state, action)
+        return state
+
+    def actions_for(self, from_state: TwentyFortyEightState, is_opponent: bool) -> List[
+            Union[TwentyFortyEightPlayerAction, TwentyFortyEightOpponentAction]]:
+        if is_opponent:
+            return [TwentyFortyEightOpponentAction(row, col, block_value=random.choices([2, 4], weights=[0.9, 0.1])[0])
+                    for row in range(4) for col in range(4)
+                    if from_state.board[row, col] == 0]
+        else:
+            return [TwentyFortyEightPlayerAction(direction) for direction in Direction if self._is_valid_move(direction, from_state)]
+
+    def _is_able_to_merge_row(self, state: TwentyFortyEightState) -> bool:
+        return any(state.board[row][col] == state.board[row][col + 1] and state.board[row][col] != 0
+                   for col in range(3) for row in range(4))
+
+    def _is_able_to_merge_col(self, state: TwentyFortyEightState) -> bool:
+        return any(state.board[row][col] == state.board[row + 1][col] and state.board[row][col] != 0
+                   for row in range(3) for col in range(4))
+
+    def _is_able_to_move_left(self, state: TwentyFortyEightState) -> bool:
+        return any(state.board[row, col] == 0 and state.board[row, col+1] != 0 for col in range(3) for row in range(4))
+
+    def _is_able_to_move_right(self, state: TwentyFortyEightState) -> bool:
+        return any(state.board[row, col] != 0 and state.board[row, col+1] == 0 for col in range(3) for row in range(4))
+
+    def _is_able_to_move_up(self, state: TwentyFortyEightState) -> bool:
+        return any(state.board[row, col] == 0 and state.board[row+1, col] != 0 for row in range(3) for col in range(4))
+
+    def _is_able_to_move_down(self, state: TwentyFortyEightState) -> bool:
+        return any(state.board[row, col] != 0 and state.board[row+1, col] == 0 for row in range(3) for col in range(4))
+
+    def _is_valid_move(self, direction: Direction, state: TwentyFortyEightState) -> bool:
+        validate_move = {
+            Direction.LEFT: (self._is_able_to_merge_row, self._is_able_to_move_left),
+            Direction.RIGHT: (self._is_able_to_merge_row, self._is_able_to_move_right),
+            Direction.UP: (self._is_able_to_merge_col, self._is_able_to_move_up),
+            Direction.DOWN: (self._is_able_to_merge_col, self._is_able_to_move_down)
+        }
+        if validate_move[direction][0](state):
+            return True
+        return validate_move[direction][1](state)
+
+    def take_action(self, state: TwentyFortyEightState,
+                    action: Union[TwentyFortyEightPlayerAction, TwentyFortyEightOpponentAction]) -> TwentyFortyEightState:
+        return action.apply(state)
+
+    def value_for_terminal(self, state: TwentyFortyEightState) -> float:
+        if 2048 in state.board:
+            return 1
+        return -1
```

Comment:
2 rzeczy:
1. zwracasz `int`, a nie `float` :P
2. może lepiej nazwać tę metodę `reward`?

Reply:
Fakt, poprawię. `reward` chyba lepiej pasuje

Reply:
Tylko wtedy trzeba było ją zmienić tutaj tak, żeby ona zwracała `0` w przypadku jeśli stan jest nieterminalny

Reply:
Hm a czemu? To miała być metoda [używana tylko dla stanów terminalnych](https://github.com/aI-lab-glider/ai-course-games-tree-search/pull/1#discussion_r758726864)

Reply:
Bo powiedziałaś, że zmienisz nazwę metody na `reward`, a `reward` sugeruje, że metoda jest nie tylko do stanów terminalnych.
Więc powiedziałem, że wtedy też trzeba dodać zabezpiszczenie, które zwracało by 0 dla stanów nieterminalnych, ponieważ nagroda dla nich nie jest zdefiniowana.

Ale jeśli nie będziemy tego zmieniać, to może też zostać tak.

Reply:
Reward to lepsza nazwa, nie wiem czemu miałaby wskazywać bardziej niż "value" na stan. To prędzej "value" w konteście Reinforcement Learning jest dla każdego stanu.

Podtrzymuję, że ta metoda powinna być używana tylko dla stanów terminalnych, ale AFAIK ciągle nigdzie nie jest to opisane (nazwa tego nie mówi / dokumentacja) <- łatwo sobie "szczelić" w stopę. Ta metoda powinna być względnie rzadko wywoływana (tylko dla stanów terminalnych), można więc rozważyć:
a) zwracanie `Optional[float]`, żeby jasno wskazać, że nie każdy stan ma wartość
b) asserta, który sypie należytym wyjątkiem ("this method should be called only on the terminal states!", jeżeli ktoś spróbuje tej metody użyć w niepoprawny sposób